src/ripple/app/tx/impl/SusPay.cpp

```diff
@@ -0,0 +1,399 @@
+        ledger entry specifies a condition, the SusPayFinish
+        must provide the sfMethod, original sfDigest, and
+        sfProof fields. Depending on the method, a
+        cryptographic operation will be performed on sfProof
+        and the result must match the sfDigest or else the
+        SusPayFinish is considered as having an invalid
+        signature.
+
+        Only sfMethod==1 is supported, where sfProof must be a
+        256-bit unsigned big-endian integer which when hashed
+        using SHA256 produces digest == sfDigest.
+
+        If the SusPay ledger entry specifies sfFinishAfter, the
+        transaction will fail if parentCloseTime <= sfFinishAfter.
+
+        SusPayFinish transactions must be submitted before a
+        SusPay's sfCancelAfter if present.
+
+        If the SusPay ledger entry specifies sfCancelAfter, the
+        transaction will fail if sfCancelAfter <= parentCloseTime.
+
+        NOTE: It must always be possible to verify the condition
+              without retrieving the SusPay ledger entry.
+
+    SusPayCancel
+
+        Any account may submit a SusPayCancel transaction.
+        
+        If the SusPay ledger entry does not specify a
+        sfCancelAfter, the cancel transaction will fail.
+
+        If parentCloseTime <= sfCancelAfter, the transaction
+        will fail.
+
+        When a SusPay is canceled, the funds are returned to
+        the source account.
+
+    By careful selection of fields in each transaction,
+    these operations may be achieved:
+
+        * Lock up XRP for a time period
+        * Execute a payment conditionally
+*/
+
+//------------------------------------------------------------------------------
+
+TER
+SusPayCreate::preflight (PreflightContext const& ctx)
+{
+    if (! (ctx.flags & tapENABLE_TESTING) &&
+            ! ctx.rules.enabled(featureSusPay))
+        return temDISABLED;
+
+    if (! isXRP(ctx.tx[sfAmount]))
+        return temBAD_AMOUNT;
+
+    if (ctx.tx[sfAmount] <= beast::zero)
+        return temBAD_AMOUNT;
+
+    if (! ctx.tx[~sfCancelAfter] &&
+            ! ctx.tx[~sfFinishAfter])
+        return temBAD_EXPIRATION;
+
+    if (ctx.tx[~sfCancelAfter] && ctx.tx[~sfFinishAfter] &&
+            ctx.tx[sfCancelAfter] <= ctx.tx[sfFinishAfter])
+        return temBAD_EXPIRATION;
+
+    return Transactor::preflight(ctx);;
+}
+
+TER
+SusPayCreate::doApply()
+{
+    // For now, require that all operations can be
+    // canceled, or finished without proof, within a
+    // reasonable period of time for the first release.
+    using namespace std::chrono;
+    auto const maxExpire =
+        ctx_.view().info().parentCloseTime +
+            seconds{days(7)}.count();
+    if (ctx_.tx[~sfDigest])
+    {
+        if (! ctx_.tx[~sfCancelAfter] ||
+                maxExpire <= ctx_.tx[sfCancelAfter])
+            return tecNO_PERMISSION;
+    }
+    else
+    {
+        if (ctx_.tx[~sfCancelAfter] &&
```

Comment:
The `~` means that the field is optional - it causes `operator[]` to return a `boost::optional` that may be `nullptr` if the field is not present, instead of generating an exception, which is what `ctx_.tx[sfCancelAfter]` would do if the field wasn't present.